### PR TITLE
[release2.1] reparo: add unit test && add safe mode

### PR DESCRIPTION
**Pull request by WangXiangUSTC**
### What problem does this PR solve? 
add unit test && add safe mode

### What is changed and how it works?
cherry pick #540 and #652 
and resolve conflicts

### Check List 

Tests 

 - Unit test
 - Integration test

Related changes
 - Need to be included in the release note

### CODE REVIEW

**Comment by suzaku** [comment]
@WangXiangUSTC 我的另一个 PR 的 Kafka 集成测试也遇到了一样的问题

**Comment by july2993** [comment]
/run-all-tests tidb=release-2.1 tikv=release-2.1 pd=release-2.1

File: tests/binlog/drainer.toml [hunk 1]
```diff
@@ -36,7 +36,7 @@ worker-count = 20
 disable-dispatch = false
 
 # safe mode will split update to delete and insert
-safe-mode = false
+safe-mode = true
```
Comment (july2993) [inline tests/binlog/drainer.toml]: why change this?

**Comment by july2993** [comment]
/run-all-tests tidb=release-2.1 tikv=release-2.1 pd=release-2.1

**Comment by WangXiangUSTC** [comment]
/run-all-tests tidb=release-2.1 tikv=release-2.1 pd=release-2.1

**Pull request merged**